src/data/polynomial/hasse_deriv.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.polynomial.derivative
+import data.nat.choose.dvd
+import data.nat.choose.vandermonde
+import logic.function.iterate
+import tactic.ring
+
+/-!
+# Hasse derivative of polynomials
+
+The `k`th Hasse derivative of a polynomial `∑ a_i X^i` is `∑ (i.choose k) a_i X^(i-k)`.
+It is a variant of the usual derivative, and satisfies `k! * (hasse_deriv k f) = derivative^[k] f`.
+The main benefit is that is gives an atomic way of talking about expressions such as
+`(derivative^[k] f).eval r / k!`, that occur in Taylor expansions, for example.
+
+## Main declarations
+
+In the following, we write `D k` for the `k`-th Hasse derivative `hasse_deriv k`.
+
+* `polynomial.hasse_deriv`: the `k`-th Hasse derivative of a polynomial
+* `polynomial.hasse_deriv_zero`: the `0`th Hasse derivative is the identity
+* `polynomial.hasse_deriv_one`: the `1`st Hasse derivative is the usual derivative
+* `polynomial.factorial_smul_hasse_deriv`: the identity `k! • (D k f) = derivative^[k] f`
+* `polynomial.hasse_deriv_comp`: the identity `(D k).comp (D l) = (k+l).choose k • D (k+l)`
+* `polynomial.hasse_deriv_mul`:
+  the "Leibniz rule" `D k (f * g) = ∑ ij in antidiagonal k, D ij.1 f * D ij.2 g`
+
+TODO: Prove the identity principle: a polynomial is 0 iff all its Hasse derivatives are zero.
+
+## Reference
+
+https://math.fontein.de/2009/08/12/the-hasse-derivative/
+
+-/
+
+noncomputable theory
+
+namespace polynomial
+
+open_locale nat big_operators
+open function nat (hiding nsmul_eq_mul)
+
+variables {R : Type*} [semiring R] (k : ℕ) (f : polynomial R)
+
+/-- The `k`th Hasse derivative of a polynomial `∑ a_i X^i` is `∑ (i.choose k) a_i X^(i-k)`.
+It satisfies `k! * (hasse_deriv k f) = derivative^[k] f`. -/
+def hasse_deriv (k : ℕ) : polynomial R →ₗ[R] polynomial R :=
+{ to_fun := λ f, f.sum $ λ i r, monomial (i - k) (↑(i.choose k) * r),
+  map_add' := λ f g,
+  begin
+    rw [sum_eq_of_subset _ _ _ _ support_add,
+        sum_eq_of_subset _ _ _ _ (f.support.subset_union_left g.support),
+        sum_eq_of_subset _ _ _ _ (f.support.subset_union_right g.support),
+        ← finset.sum_add_distrib, finset.sum_congr rfl],
+    { intros i hi, simp only [coeff_add, mul_add, linear_map.map_add], },
+    all_goals { simp only [forall_const, monomial_zero_right, mul_zero], },
+  end,
```

Comment:
I feel like it should be possible to use `finsupp.lsum` to avoid needing to prove this manually. @urkud, do you know if we have a polynomial spelling of lsum?

Reply:
golfed